File: test/test_java_project/Dog.java
package animals;

import java.util.ArrayList;
import java.util.List;

/**
 * Dog class demonstrating inheritance and polymorphism.
 */
public class Dog extends Animal {
    private String breed;
    private List<String> tricks;

    public Dog(String name, int age, String breed) {
        super(name, age);
        this.breed = breed;
        this.tricks = new ArrayList<>();
    }

    @Override
    public void speak() {
        System.out.println(getName() + " says: Woof!");
    }

    @Override
    public void move() {
        System.out.println(getName() + " runs on four legs");
    }

    public void fetch() {
        System.out.println(getName() + " fetches the ball!");
        learnTrick("fetch");
    }

    public void learnTrick(String trick) {
        tricks.add(trick);
        System.out.println(getName() + " learned: " + trick);
    }

    public List<String> getTricks() {
        return new ArrayList<>(tricks);
    }

    public String getBreed() {
        return breed;
    }

    // Nested loop for complexity testing
    public void performAllTricks(int times) {
        for (int i = 0; i < times; i++) {
            for (String trick : tricks) {
                System.out.println(getName() + " performs: " + trick);
            }
        }
    }
}

File: test/test_c_project/dog.c
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include "dog.h"

/* Override Animal methods for Dog */
static void dog_speak(Animal* self) {
    Dog* dog = (Dog*)self;
    printf("%s barks: Woof! (breed: %s)\n", self->name, dog->breed);
}

static void dog_move(Animal* self) {
    printf("%s runs on four legs\n", self->name);
}

static void dog_destroy_impl(Animal* self) {
    Dog* dog = (Dog*)self;
    free(dog->breed);
    free(self->name);
    free(dog);
}

/* Dog's vtable - polymorphism */
static const AnimalVtable dog_vtable = {
    .speak = dog_speak,
    .move = dog_move,
    .destroy = dog_destroy_impl
};

Dog* dog_new(const char* name, int age, const char* breed) {
    Dog* d = malloc(sizeof(Dog));
    if (!d) return NULL;

    d->base.vtable = &dog_vtable;
    d->base.name = strdup(name);
    d->base.age = age;
    d->breed = strdup(breed);
    d->is_trained = 0;
    return d;
}

void dog_free(Dog* self) {
    if (self) {
        self->base.vtable->destroy((Animal*)self);
    }
}

void dog_bark(Dog* self) {
    for (int i = 0; i < 3; i++) {
        printf("Woof! ");
    }
    printf("\n");
}

void dog_fetch(Dog* self) {
    printf("%s fetches the ball!\n", self->base.name);
    self->is_trained = 1;
}
